hswidget/cof_widget.go

```diff
@@ -0,0 +1,660 @@
+package hswidget
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/ianling/giu"
+	"github.com/ianling/imgui-go"
+
+	"github.com/OpenDiablo2/OpenDiablo2/d2common/d2enum"
+	"github.com/OpenDiablo2/OpenDiablo2/d2common/d2fileformats/d2cof"
+
+	"github.com/OpenDiablo2/HellSpawner/hscommon/hsenum"
+)
+
+const (
+	indicatorSize                        = 64
+	upDownArrowW, upDownArrowH           = 15, 15
+	leftRightArrowW, leftRightArrowH     = 15, 15
+	actionButtonW, actionButtonH         = 200, 30
+	saveCancelButtonW, saveCancelButtonH = 80, 30
+	bigListW                             = 200
+	trueFalseListW                       = 60
+	speedInputW                          = 40
+)
+
+type cofTextures struct {
+	up    *giu.Texture
+	down  *giu.Texture
+	left  *giu.Texture
+	right *giu.Texture
+}
+
+// COFWidget represents cof viewer's widget
+type COFWidget struct {
+	id       string
+	cof      *d2cof.COF
+	textures cofTextures
+}
+
+// COFViewer creates a cof viewer widget
+func COFViewer(
+	up, down, right, left *giu.Texture,
+	id string, cof *d2cof.COF,
+) *COFWidget {
+	result := &COFWidget{
+		id:  id,
+		cof: cof,
+	}
+
+	result.textures.up = up
+	result.textures.down = down
+	result.textures.left = left
+	result.textures.right = right
+
+	return result
+}
+
+// Build builds a cof viewer
+func (p *COFWidget) Build() {
+	stateID := fmt.Sprintf("COFWidget_%s", p.id)
+	s := giu.Context.GetState(stateID)
+
+	if s == nil {
+		p.setDefaultState(stateID)
+		return
+	}
+
+	state := s.(*COFState)
```

Comment:
hmm. maybe it could be cut like in another widgets by `p.getState()` method?

Reply:
you mean `getState` method should handle setting up the default state on first retrieval or something?

Reply:
resolved in ed7ceff